napari/_qt/_tests/test_plugin_widgets.py

```diff
@@ -195,3 +133,65 @@ def test_inject_viewer_proxy(make_napari_viewer):
     with patch('napari.utils.misc.ROOT_DIR', new='/some/other/package'):
         with pytest.warns(FutureWarning):
             wdg.fail()
+
+
+@pytest.mark.parametrize(
+    "widget_callable, param",
+    [
+        (QWidget_example, 'napari_viewer'),
+        (QWidget_string_annnot, 'test'),
+        (Container_example, 'test'),
+    ],
+)
+def test_get_widget_viewer_param(widget_callable, param):
+    """Test `_get_widget_viewer_param` returns correct parameter name."""
+    out = _get_widget_viewer_param(widget_callable, 'widget_name')
+    assert out == param
+
+
+def test_get_widget_viewer_param_error():
+    """Test incorrect subclass raises error in `_get_widget_viewer_param`."""
+    with pytest.raises(TypeError) as e:
+        _get_widget_viewer_param(ErrorWidget, 'widget_name')
+    assert "'widget_name' must be `QtWidgets.QWidget`" in str(e)
+
+
+def test_widget_hide_destroy(make_napari_viewer):
+    """Test that widget hide and destroy works."""
+    viewer = make_napari_viewer()
+    viewer.window.add_dock_widget(QWidget_example(), name='test')
+    widget = viewer.window._dock_widgets['test']
+
+    # Check widget persists after hide
+    widget.title.hide_button.click()
+    assert widget
+    # Check that widget removed from `_dock_widgets` dict when closed
+    widget.destroyOnClose()
+    assert 'test' not in viewer.window._dock_widgets
+
+
+@pytest.mark.parametrize(
+    "Widget",
+    [
+        QWidget_example,
+        Container_example,
+        magic_widget_example,
+        callable_example,
+    ],
+)
+def test_widget_types_supported(
+    make_napari_viewer, tmp_plugin: DynamicPlugin, Widget
+):
+    """Test all supported widget types correctly instantiated and call processor."""
+
+    @tmp_plugin.contribute.widget(display_name='Widget')
+    def widget():
+        return Widget()
+
+    app = get_app()
+    # `strict_qt` prevents checking for leaked widgets resulting from parametrize
+    viewer = make_napari_viewer()
+
+    viewer.window.add_dock_widget = MagicMock()
```

Comment:
```suggestion
    viewer.window.add_dock_widget = Mock(side_effect=viewer.window.add_dock_widget)
```
This seems to prevent widgets from being leaked for me.